torch/_higher_order_ops/scan.py

```diff
@@ -0,0 +1,339 @@
+            )
+
+            # Store the inits in the outs matrix.
+            store_out_in_outs(out, ind + 1)
+
+            ind += 1
+
+        return outs
+
+    if len(input) == 0:
+        return []
+
+    scans = _scan(input, init)
+    return scans
+
+
+def trace_scan(
+    proxy_mode,
+    func_overload,
+    combine_fn: Callable,
+    input: List[torch.Tensor],
+    init: List[torch.Tensor],
+    dim: int,
+):
+    with disable_proxy_modes_tracing():
+        sample_inputs = [
+            torch.empty_like(
+                x_init,
+                dtype=x.dtype,
+                device=x.device,
+                requires_grad=x.requires_grad,
+            )
+            for x, x_init in itertools.chain(zip(input, init), zip(input, init))
+        ]
+        combine_graph = reenter_make_fx(combine_fn)(*sample_inputs)
+
+    outputs = None
+    for node in combine_graph.graph.nodes:
+        if node.op == "output":
+            assert outputs is None
+            assert len(node.args) == 1
+            outputs = node.args[0]
+
+    assert outputs is not None
+    assert len(outputs) == len(
+        input
+    ), f"expected combine_fn to return {len(input)} results but got {len(outputs)}"
+
+    for i, si, o in zip(input, sample_inputs, outputs):
+        o_meta = o.meta["tensor_meta"]
+        assert o_meta.dtype == i.dtype, (
+            f"combine_fn output type mismatch, expected {i.dtype} "
+            + f"but got {o_meta.dtype}"
+        )
+        assert (
+            si.shape == o_meta.shape
+        ), "The pytree of the out of the operator needs to match the input pytree"
+
+    _, combine_graph_name = unique_graph_id(proxy_mode, prefix="scan_combine_graph")
+
+    proxy_mode.tracer.root.register_module(combine_graph_name, combine_graph)
+
+    args = (combine_graph, input, dim)
+    proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, args)
+    out_proxy = proxy_mode.tracer.create_proxy(
+        "call_function", func_overload, proxy_args, {}, name="scan"
+    )
+
+    with disable_proxy_modes_tracing():
+        out = [aten.clone(x) for x in input]
+
+    return track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)
+
+
+@scan_op.py_impl(DispatchKey.CompositeExplicitAutograd)
+def scan_op_dense(combine_fn, input, init, dim):
+    mode = _get_current_dispatch_mode()
+    assert mode is None, "Mode should never be enabled for CPU/CUDA key"
+    return generic_scan(combine_fn, input, init, dim)
+
+
+scan_op.py_impl(DispatchKey.Autograd)(
+    autograd_not_implemented(scan_op, deferred_error=True)
+)
+
+
+@scan_op.py_impl(ProxyTorchDispatchMode)
+def scan_proxy_mode(mode, combine_fn, input, init, dim):
+    return trace_scan(mode, scan_op, combine_fn, input, init, dim)
+
+
+@scan_op.py_impl(FakeTensorMode)
+def assoiciative_scan_fake_tensor_mode(mode, combine_fn, input, init, dim):
+    with mode:
+        return combine_fn(*input, *input)
```

Comment:
This is not correct because the fake output of scan should be the stacked output of combine_fn.

Reply:
I corrected this part. Interesting there was no testcase failing because of that. I also added a [dedicated test](https://github.com/pytorch/pytorch/blob/810491fddfb4e51cdbe4e56c7094c8a56729c5bd/test/test_fake_tensor.py#L928) for the FakeTensorMode in test_fake_tensor.py.